core/vm/evmc.go

```diff
@@ -0,0 +1,376 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/evmc/bindings/go/evmc"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type EVMC struct {
+	instance *evmc.Instance
+	env      *EVM
+	readOnly bool // TODO: The readOnly flag should not be here.
+}
+
+var (
+	createMu     sync.Mutex
+	evmcConfig   string // The configuration the instance was created with.
+	evmcInstance *evmc.Instance
+)
+
+func createVM(config string) *evmc.Instance {
+	createMu.Lock()
+	defer createMu.Unlock()
+
+	if evmcInstance == nil {
+		options := strings.Split(config, ",")
+		path := options[0]
+
+		if path == "" {
+			panic("EVMC VM path not provided, set --vm.(evm|ewasm)=/path/to/vm")
+		}
+
+		var err error
+		evmcInstance, err = evmc.Load(path)
+		if err != nil {
+			panic(err.Error())
+		}
+		log.Info("EVMC VM loaded", "name", evmcInstance.Name(), "version", evmcInstance.Version(), "path", path)
+
+		for _, option := range options[1:] {
+			if idx := strings.Index(option, "="); idx >= 0 {
+				name := option[:idx]
+				value := option[idx+1:]
+				err := evmcInstance.SetOption(name, value)
+				if err == nil {
+					log.Info("EVMC VM option set", "name", name, "value", value)
+				} else {
+					log.Warn("EVMC VM option setting failed", "name", name, "error", err)
+				}
+			}
+		}
+
+		evm1Cap := evmcInstance.HasCapability(evmc.CapabilityEVM1)
+		ewasmCap := evmcInstance.HasCapability(evmc.CapabilityEWASM)
+		log.Info("EVMC VM capabilities", "evm1", evm1Cap, "ewasm", ewasmCap)
+
+		evmcConfig = config // Remember the config.
+	} else if evmcConfig != config {
+		log.Error("New EVMC VM requested", "newconfig", config, "oldconfig", evmcConfig)
+	}
+	return evmcInstance
+}
+
+func NewEVMC(options string, env *EVM) *EVMC {
+	return &EVMC{createVM(options), env, false}
```

Comment:
Let's inline createVM here, and also return an `error` too. Makes handling "bad flags" cleaner than logging warnings.